File: tests/test_discovery_registry.rs
use std::borrow::Cow;

use openchecks::{
    AsyncCheck, Check, CheckHint, CheckMetadata, CheckResult, DiscoveryRegistry, Item,
};

#[derive(Debug, Clone, Copy, PartialEq, PartialOrd)]
struct TestItem {
    value: u8,
}

impl std::fmt::Display for TestItem {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.value)
    }
}

impl Item for TestItem {
    type Value<'a>
        = u8
    where
        Self: 'a;

    fn value(&self) -> Self::Value<'_> {
        self.value
    }
}

struct TestCheck;

impl CheckMetadata for TestCheck {
    fn title(&self) -> Cow<str> {
        "TestCheck".into()
    }

    fn description(&self) -> Cow<str> {
        "Description".into()
    }

    fn hint(&self) -> CheckHint {
        CheckHint::NONE
    }
}

impl Check for TestCheck {
    type Item = TestItem;
    type Items = Vec<Self::Item>;

    fn check(&self) -> openchecks::CheckResult<Self::Item, Self::Items> {
        CheckResult::new_passed("message", None, false, false)
    }
}

#[async_trait::async_trait]
impl AsyncCheck for TestCheck {
    type Item = TestItem;
    type Items = Vec<Self::Item>;

    async fn async_check(&self) -> openchecks::CheckResult<Self::Item, Self::Items> {
        CheckResult::new_passed("message", None, false, false)
    }
}

#[test]
fn test_discovery_registry_register_and_gather_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register(|_ctx: &()| true, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather(&());

    assert!(result.is_some());
    assert_eq!(result.unwrap().len(), 1);
}

#[test]
fn test_discovery_registry_gather_empty_plugins_success() {
    let registry: DiscoveryRegistry<(), TestItem, Vec<_>> = DiscoveryRegistry::new();
    let result = registry.gather(&());

    assert!(result.is_none());
}

#[test]
fn test_discovery_registry_query_context_gather_return_some_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register(|_ctx: &()| true, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather(&());

    assert!(result.is_some());
}

#[test]
fn test_discovery_registry_query_context_gather_return_none_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register(|_ctx: &()| false, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather(&());

    assert!(result.is_none());
}

#[test]
fn test_discovery_registry_register_and_gather_async_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register_async(|_ctx: &()| true, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather_async(&());

    assert!(result.is_some());
    assert_eq!(result.unwrap().len(), 1);
}

#[test]
fn test_discovery_registry_gather_async_empty_plugins_success() {
    let registry: DiscoveryRegistry<(), TestItem, Vec<_>> = DiscoveryRegistry::new();
    let result = registry.gather_async(&());

    assert!(result.is_none());
}

#[test]
fn test_discovery_registry_query_context_gather_async_return_some_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register_async(|_ctx: &()| true, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather_async(&());

    assert!(result.is_some());
}

#[test]
fn test_discovery_registry_query_context_gather_async_return_none_success() {
    let mut registry = DiscoveryRegistry::new();
    registry.register_async(|_ctx: &()| false, |_ctx| vec![Box::new(TestCheck)]);
    let result = registry.gather_async(&());

    assert!(result.is_none());
}

File: bindings/python/tests/test_discovery_registry.py
from __future__ import annotations

import pytest

import openchecks


class Check(openchecks.BaseCheck):
    def check(self) -> openchecks.CheckResult[int]:
        return openchecks.CheckResult.passed("test")

    def title(self) -> str:
        return "title"

    def description(self) -> str:
        return "description"


class AsyncCheck(openchecks.AsyncBaseCheck):
    async def async_check(self) -> openchecks.CheckResult[int]:
        return openchecks.CheckResult.passed("test")

    def title(self) -> str:
        return "title"

    def description(self) -> str:
        return "description"


def test_discovery_registry_register_and_gather_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    registry.register(lambda _: True, lambda _: [Check()])
    result = registry.gather(None)

    assert result
    assert len(result) == 1


def test_discovery_registry_gather_empty_plugins_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    result = registry.gather(None)

    assert result is None


def test_discovery_registry_query_context_gather_return_some_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    registry.register(lambda _: True, lambda _: [Check()])
    result = registry.gather(None)

    assert result is not None


def test_discovery_registry_query_context_gather_return_none_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    registry.register(lambda _: False, lambda _: [Check()])
    result = registry.gather(None)

    assert result is None


@pytest.mark.asyncio
async def test_discovery_registry_register_and_gather_async_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    registry.register_async(lambda _: True, lambda _: [AsyncCheck()])
    result = registry.gather_async(None)

    assert result
    assert len(result) == 1


@pytest.mark.asyncio
async def test_discovery_registry_gather_async_empty_plugins_success() -> None:
    registry = openchecks.DiscoveryRegistry()
    result = registry.gather_async(None)

    assert result is None


@pytest.mark.asyncio
async def test_discovery_registry_query_context_gather_async_return_some_success() -> (
    None
):
    registry = openchecks.DiscoveryRegistry()
    registry.register_async(lambda _: True, lambda _: [AsyncCheck()])
    result = registry.gather_async(None)

    assert result


@pytest.mark.asyncio
async def test_discovery_registry_query_context_gather_async_return_none_success() -> (
    None
):
    registry = openchecks.DiscoveryRegistry()
    registry.register_async(lambda _: False, lambda _: [AsyncCheck()])
    result = registry.gather_async(None)

    assert result is None
